specs/Proposals/0040-split-states.md

```diff
@@ -0,0 +1,257 @@
+
+Whether a receipt is processed in a block or delayed can affect the execution result of this receipt because transactions are charged and local receipts are processed before delayed receipts are processed. For example, let’s assume Alice’s account has 0N now and Bob sends a transaction T1 to transfer 5N to Alice. The transaction has been converted to a receipt R at block i-1 and sent to Alice's shard at block i. Let's say Alice signs another transaction T2 to send 1N to Charlie and that transaction is included in block i+1. Whether transaction T2 succeeds depends on whether receipt R is processed or delayed in block i. If R is processed in block i, Alice’s account will have 5N before block i+1 and T2 will succeed while if R is delayed in block i, Alice’s account will have 0N and T2 will be declined.
+
+Therefore, the validators still process transactions and receipts based on the current sharding assignment. After the processing is finished, they can take the generated state changes to apply to the new states.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+## Sharding Representation
+### `EpochManager`
+`EpochManager` will be responsible for managing shards info accross epochs.
+#### `from_shard_id_to_ord`
+```rust
+from_shard_id_to_ord(&self, shard_id: ShardId, current_epoch_id: &EpochId) -> Result<(EpochId, ShardOrd), EpochError>
+```
+converts `shard_id` to the `shard_ord` and `epoch_id` of a shard. `current_epoch_id` will be used as the starting search point, the function will only search for epoch T-1, T, T+1 if the current epoch is T. If it cannot find the `shard_id`, it returns an `EpochError`.
+#### `from_shard_ord_to_id`
+```rust
+from_shard_ord_to_id(&self, shard_ord: ShardOrd, epoch_id: &EpochId) -> Result<ShardId, EpochError>
+```
+converts `shard_ord` and `epoch_id` for a shard to its `shard_id`. It returns an EpochError if such shard does not exist.
+### EpochInfoV3
+Epoch info will include the `shard_id` of the shards in this epoch and a mappting from the current shards to their parent shards. We only allow adding new shards that are split from the existing shards, and if shard B and C are split from shard A, we call shard A the parent shard of shard B and C. For example, if epoch T-1 has shard with `shard_id` 0 and 1 and each of them will be split to two shards in epoch T, then the `EpochInfo` for epoch T will have `shards = [2, 3, 4, 5]` and `parent_shards = {2:0, 3:0, 4:1, 5:1}`.
+
+```rust
+pub struct EpochInfoV3 {
+    ... // All fields in EpochInfoV2
+    shards: Vec<ShardId>,
+    parent_shards: HashMap<ShardId, ShardId>,
+}
+```
+
+### ShardVersionManager
+A new struct `ShardVersionManager` will be created to manage sharding assignment when `ProtocolVersion` is changed. `EpochManager` will own `ShardVersionManager` that it can call to create shards for the `EpochInfo` of the next next epoch.
+
```

Comment:
Given that this information is only needed to access state, I am not sure whether epoch manager is the best place to store it.

Reply:
I decided to put it inside EpochManager because shards information are saved as part of EpochInfo. EpochManager manages protocol version of epochs, and in this design, shard version is derived there, so it seems natural to put this information there.

Reply:
discussed offline